### PR TITLE
Feature/dynamic types

### PR DESCRIPTION
**Pull request by LuisGP**
Changes related to dynamic types and bitset support.

### CODE REVIEW

File: src/java/com/eprosima/fastcdr/idl/templates/FastCdrCommon.stg [hunk 1]
```diff
@@ -68,54 +68,106 @@ $endif$
...
+
+$if(typecode.isWCharType)$
+$var$ += 4$if(ctx.cdr)$ + eprosima::fastcdr::Cdr::alignment($var$, 4)$endif$; // 32 bits
+$elseif(typecode.isType_13)$
+$var$ += 16$if(ctx.cdr)$ + eprosima::fastcdr::Cdr::alignment($var$, 8)$endif$; // 128 bits
+$elseif(typecode.primitive)$
 $var$ += $typecode.size$$if(ctx.cdr)$ + eprosima::fastcdr::Cdr::alignment($var$, $typecode.size$)$endif$;
+
+$elseif(typecode.isWStringType)$
+$var$ += 4$if(ctx.cdr)$ + eprosima::fastcdr::Cdr::alignment($var$, 4)$endif$ + ($typecode.maxsize$) * 4; // 32 bits
```
[the remark this answers is not in the document]
Comment (LuisGP) [reply src/java/com/eprosima/fastcdr/idl/templates/FastCdrCommon.stg]: In section 7.4.3.5.3 (page 109 and 110) of the Extensible and Dynamic Topic Types for DDS, it is established that wstring doesn't include the trailing zero, in contrast of string.

File: src/java/com/eprosima/fastcdr/idl/templates/FastCdrCommon.stg [hunk 2]
```diff
@@ -68,54 +68,106 @@ $endif$
...
+
+$if(object.typecode.isWCharType)$
+$var$ += 4$if(ctx.cdr)$ + eprosima::fastcdr::Cdr::alignment($var$, 4)$endif$; // 32 bits
+$elseif(object.typecode.isType_13)$
+$var$ += 16$if(ctx.cdr)$ + eprosima::fastcdr::Cdr::alignment($var$, 8)$endif$; // 128 bits
+$elseif(object.typecode.primitive)$
 $var$ += $object.typecode.size$$if(ctx.cdr)$ + eprosima::fastcdr::Cdr::alignment($var$, $object.typecode.size$)$endif$;
+
+$elseif(object.typecode.isWStringType)$
+$var$ += 4$if(ctx.cdr)$ + eprosima::fastcdr::Cdr::alignment($var$, 4)$endif$ + ($data$.$object.name$().size()) * 4; // 32 bits
```
[the remark this answers is not in the document]
Comment (LuisGP) [reply src/java/com/eprosima/fastcdr/idl/templates/FastCdrCommon.stg]: In section 7.4.3.5.3 (page 109 and 110) of the Extensible and Dynamic Topic Types for DDS, it is established that wstring doesn't include the trailing zero, in contrast of string.

File: src/java/com/eprosima/fastcdr/idl/templates/FastCdrCommon.stg [hunk 3]
```diff
@@ -68,54 +68,106 @@ $endif$
...
 $var$ += $typecode.scopedname$::getMaxCdrSerializedSize($var$);
 $endif$
+
 >>
 
 serialized_size_type(ctx, typecode, var) ::= <<
-$if(typecode.primitive)$
+/* $typecode.cppTypename$ */
+$if(typecode.isWCharType)$
+$var$ += 4$if(ctx.cdr)$ + eprosima::fastcdr::Cdr::alignment($var$, 4)$endif$; // 32 bits
```
[the remark this answers is not in the document]
Comment (LuisGP) [reply src/java/com/eprosima/fastcdr/idl/templates/FastCdrCommon.stg]: In wchar case you are right, but long double is aligned with 8 bytes instead 16, so the primitive behaviour doesn't suit for it. I'll modify it to show this.

File: src/java/com/eprosima/fastcdr/idl/templates/FastCdrCommon.stg [hunk 4]
```diff
@@ -68,54 +68,106 @@ $endif$
...
 serialized_size(ctx, object, data, var) ::= <<
-$if(object.typecode.primitive)$
+$if(object.typecode.isMapType)$
+/* map<$object.typecode.keyTypeCode.cppTypename$, $object.typecode.valueTypeCode.cppTypename$, $object.typecode.maxsize$> $object.name$ */
+$else$
+/* $object.typecode.cppTypename$ $object.name$ */
+$endif$
+
+$if(object.typecode.isWCharType)$
+$var$ += 4$if(ctx.cdr)$ + eprosima::fastcdr::Cdr::alignment($var$, 4)$endif$; // 32 bits
```
[the remark this answers is not in the document]
Comment (LuisGP) [reply src/java/com/eprosima/fastcdr/idl/templates/FastCdrCommon.stg]: In wchar case you are right, but long double is aligned with 8 bytes instead 16, so the primitive behaviour doesn't suit for it. I'll modify it to show this.

File: src/java/com/eprosima/fastcdr/idl/templates/FastCdrCommon.stg [hunk 5]
```diff
@@ -68,54 +68,106 @@ $endif$
...
 max_serialized_size(ctx, typecode, var) ::= <<
-$if(typecode.primitive)$
+$if(typecode.isMapType)$
+/* map<$typecode.keyTypeCode.cppTypename$, $typecode.valueTypeCode.cppTypename$, $typecode.maxsize$> */
+$else$
+/* $typecode.cppTypename$ */
+$endif$
+
+$if(typecode.isWCharType)$
+$var$ += 4$if(ctx.cdr)$ + eprosima::fastcdr::Cdr::alignment($var$, 4)$endif$; // 32 bits
```
[the remark this answers is not in the document]
Comment (LuisGP) [reply src/java/com/eprosima/fastcdr/idl/templates/FastCdrCommon.stg]: In wchar case you are right, but long double is aligned with 8 bytes instead 16, so the primitive behaviour doesn't suit for it. I'll modify it to show this.

File: src/cpp/Cdr.cpp [hunk 6]
```diff
@@ -1571,14 +1689,59 @@ const char* Cdr::readString(uint32_t &length)
...
+        wchar_t* wValue = new wchar_t[length];
+        deserializeArray(wValue, length);
+        returnedValue = std::wstring(wValue, length);
+        delete [] wValue;
+#else
+        const wchar_t* wValue = L"";
+        wValue = reinterpret_cast<wchar_t*>(&m_currentPosition);
+        m_currentPosition += bytesLength;
+#endif
+        if (wValue[length - 1] == '\0')
```
[the remark this answers is not in the document]
Comment (richiware) [reply src/cpp/Cdr.cpp]: Still wrong. On windows is set `returnedValue` twice.

File: src/java/com/eprosima/fastcdr/idl/templates/FastCdrCommon.stg [hunk 7]
```diff
@@ -196,6 +270,69 @@ $endif$
 $endif$
 >>
 
+map_max_serialized_size(ctx, typecodeMap, var, loopvar) ::= <<
+for(size_t $loopvar$ = 0; $loopvar$ < $typecodeMap.maxsize$; ++$loopvar$)
```
[the remark this answers is not in the document]
Comment (richiware) [reply src/java/com/eprosima/fastcdr/idl/templates/FastCdrCommon.stg]: Ok, I've annotated the improve in these generated codes.

**Comment by richiware** [comment]
CI failed in [#1](http://ci.intranet.eprosima.com:8080/view/Manual/job/FastCDR%20Manual/1/) and please, rebase with develop.

**Comment by richiware** [comment]
Failed CI [#5](http://ci.intranet.eprosima.com:8080/job/FastCDR%20Manual%20Linux/5/). Valgrind detected something. Check.

**Pull request merged**